src/pcre2_internal.h

```diff
@@ -2042,6 +2042,38 @@ extern void *       _pcre2_memmove(void *, const void *, size_t);
 #endif
 
 #endif  /* PCRE2_CODE_UNIT_WIDTH */
+
+/*************************************************
+*        Checked Integer Multiplication          *
+*************************************************/
+
+/*
+Arguments:
+  r         A pointer to PCRE2_SIZE to store the answer
+  a, b      Two positive integers
+
+Returns:    Bool indicating if the operation overflows
+
+It is modeled after C23's <stdckdint.h> interface
+The INT64_OR_DOUBLE type is a 64-bit integer type when available,
+otherwise double. */
+
+static inline BOOL
+PRIV(ckd_smul)(PCRE2_SIZE *r, int a, int b)
+{
+INT64_OR_DOUBLE m;
+
+#ifndef NDEBUG
+if (a < 0 || b < 0) abort();
+#endif
+
+m = (INT64_OR_DOUBLE)a * (INT64_OR_DOUBLE)b;
+if (!(PCRE2_SIZE_MAX - m > 0)) return TRUE;
```

Comment:
`m < PCRE2_SIZE_MAX` is enough
btw if either argument is negative this is true. You could use unsigned multiplication if negative numbers are not possible. 
I also remember something when int max is converted to infinity, but it might be the opposite conversion direction.

Reply:
That is how I wrote it [originally](https://github.com/PCRE2Project/pcre2/actions/runs/4006442948/jobs/6877970854), but it triggered unhelpful (-Wsign-compare) warnings in most builds.

Note that using unsigned here isn't useful as the original variables are signed, and indeed this whole function exists solely to prevent triggering undefined behaviour with signed integer multiplications that overflow.